datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -502,6 +589,140 @@ mod tests {
         });
     }
 
+    #[test]
+    fn test_date_bin_single() {
+        use chrono::Duration;
+
+        let cases = vec![
+            (
+                (
+                    Duration::minutes(15),
+                    "2004-04-09T02:03:04.123456789Z",
+                    "2001-01-01T00:00:00",
+                ),
+                "2004-04-09T02:00:00Z",
+            ),
+            (
+                (
+                    Duration::minutes(15),
+                    "2004-04-09T02:03:04.123456789Z",
+                    "2001-01-01T00:02:30",
+                ),
+                "2004-04-09T02:02:30Z",
+            ),
+            (
+                (
+                    Duration::minutes(15),
+                    "2004-04-09T02:03:04.123456789Z",
+                    "2005-01-01T00:02:30",
+                ),
+                "2004-04-09T02:02:30Z",
+            ),
+        ];
+
+        cases
+            .iter()
+            .for_each(|((stride, source, origin), expected)| {
+                let stride1 = stride.num_nanoseconds().unwrap();
+                let source1 = string_to_timestamp_nanos(source).unwrap();
+                let origin1 = string_to_timestamp_nanos(origin).unwrap();
+
+                let expected1 = string_to_timestamp_nanos(expected).unwrap();
+                let result = date_bin_single(stride1, source1, origin1);
+                assert_eq!(result, expected1, "{} = {}", source, expected);
+            })
+    }
+
+    #[test]
+    fn test_date_bin() {
+        let res = date_bin(&[
+            ColumnarValue::Scalar(ScalarValue::IntervalDayTime(Some(1))),
+            ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(Some(1), None)),
+            ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(Some(1), None)),
+        ]);
+        assert!(res.is_ok());
+
+        let mut builder = TimestampNanosecondArray::builder(5);
+        builder.append_slice((1..6).collect::<Vec<i64>>().as_slice());
+        let timestamps = Arc::new(builder.finish());
```

Comment:
It doesn't matter for correctness, but here is an alternative potential construction if you want to use a more functional style

```suggestion
        let timestamps = Arc::new(
           (1..6).map(Some)
           .collect::<TimestampNanosecondArray>()
          );
```

Reply:
Much nicer – thanks!